lib/logstash/codecs/identity_map_codec.rb

```diff
@@ -0,0 +1,176 @@
+# encoding: utf-8
+require "logstash/namespace"
+require "concurrent"
+
+# This class is a Codec duck type
+# Using Composition, it maps from a stream identity to
+# a cloned codec instance via the same API as a Codec
+
+module LogStash module Codecs class IdentityMapCodec
+  # subclass of Exception, LS has more than limit (100) active streams
+  class IdentityMapUpperLimitException < Exception; end
+
+  module EightyPercentWarning
+    extend self
+    def visit(imc)
+      current_size, limit = imc.current_size_and_limit
+      return if current_size < (limit * 0.8)
+      imc.logger.warn("IdentityMapCodec has reached 80% capacity",
+        :current_size => current_size, :upper_limit => limit)
+    end
+  end
+
+  module UpperLimitReached
+    extend self
+    def visit(imc)
+      current_size, limit = imc.current_size_and_limit
+      return if current_size < limit
+      current_size, limit = imc.map_cleanup
+      return if current_size < limit
+      imc.logger.error("IdentityMapCodec has reached 100% capacity",
+          :current_size => current_size, :upper_limit => limit)
+      raise IdentityMapUpperLimitException.new
+    end
+  end
+
+  class MapCleaner
+    def initialize(imc, interval)
+      @running = true
+      @imc, @interval = imc, interval
+    end
+
+    def run
+      @thread = Thread.new(@imc) do |imc|
+        loop do
+          sleep @interval
+          break if !@running
+          imc.map_cleanup
+        end
+      end
+      self
+    end
+
+    def stop
+      @running = false
+      @thread.wakeup
+    end
+  end
+
+  MAX_IDENTITIES = 100
+  EVICT_TIMEOUT = 60 * 60 * 4 # 4 hours
+  CLEANER_INTERVAL = 60 * 5 # 5 minutes
+
+  attr_reader :identity_map, :usage_map
+  attr_accessor :base_codec, :logger, :cleaner
+
+  def initialize(codec, logger)
+    @base_codec = codec
+    @base_values = [codec]
+    @identity_map = Hash.new &method(:codec_builder)
+    # @identity_map = Concurrent::Hash.new &method(:codec_builder)
+    @usage_map = Hash.new
+    # @usage_map = Concurrent::Hash.new
+    @logger = logger
+    @max_identities = MAX_IDENTITIES
+    @evict_timeout = EVICT_TIMEOUT
+    @cleaner = MapCleaner.new(self, CLEANER_INTERVAL).run
+  end
+
+  def max_identities(max)
+    @max_identities = max.to_i.abs
+    self
+  end
+
+  def evict_timeout(timeout)
+    @evict_timeout = timeout.to_i.abs
+    self
+  end
+
+  def cleaner_interval(interval)
```

Comment:
ok. AFAIK *fluent* style has not been used in the logstash code base. I am not against it but we should probably discuss if we want to introduce that style. In any way, I suggest you document it is the class definition with probably an example so that eventual plugin authors that want to use this will know how to use this class?